### PR TITLE
[serve] Make serve agent not blocking when GCS is down.

### PR DESCRIPTION
**Pull request by fishbone**
Signed-off-by: Yi Cheng <chengyidna@gmail.com>

## Why are these changes needed?

This PR fixed several issue which block serve agent when GCS is down. We need to make sure serve agent is always alive and can make sure the external requests can be sent to the agent and check the status.

- internal kv used in dashboard/agent blocks the agent. We use the async one instead
- serve controller use ray.nodes which is a blocking call and blocking forever. change to use gcs client with timeout
- agent use serve controller client which is a blocking call with max retries = -1. This blocks until controller is back.

To enable Serve HA, we also need to setup:

- RAY_gcs_server_request_timeout_seconds=5
- RAY_SERVE_KV_TIMEOUT_S=5

which we should set in KubeRay.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by brucez-anyscale — COMMENTED** [review]
lgtm, I am not that familiar. Will let others stamp it.

File: dashboard/modules/serve/serve_agent.py
Comment (edoakes) [inline dashboard/modules/serve/serve_agent.py:62]: hmm I think we should raise an exception here instead of returning the empty schema -- that is incorrect behavior (there may actually be an active config)

cc @shrekris-anyscale

File: python/ray/serve/_private/http_state.py
Comment (edoakes) [inline python/ray/serve/_private/http_state.py:55]: why delay this?

File: dashboard/modules/serve/serve_agent.py
Comment (shrekris-anyscale) [reply dashboard/modules/serve/serve_agent.py:62]: Agreed– the empty schema indicates that there's no active config. We need to communicate that the GCS is currently recovering in this exception.

Comment (iycheng) [reply dashboard/modules/serve/serve_agent.py:62]: @edoakes @shrekris-anyscale   I need to sync with @brucez-anyscale about what serve operator wants for this case. I'll update it and the test early in the afternoon.

File: python/ray/serve/_private/http_state.py
Comment (iycheng) [reply python/ray/serve/_private/http_state.py:55]: There is one test (https://github.com/ray-project/ray/blob/HEAD/python/ray/serve/tests/test_http_state.py) running without ray, so the connection will fail.

Do you think I should just make it run with ray?

Comment (edoakes) [reply python/ray/serve/_private/http_state.py:55]: Ah I see, can we pass the gcs_client into `HTTPState` and in this unit test pass in a mock?

File: dashboard/modules/serve/serve_agent.py
Comment (edoakes) [reply dashboard/modules/serve/serve_agent.py:62]: Ok sounds good. I think the most intuitive behavior would be to return a `503 Service Unavailable` in these cases.

File: dashboard/modules/serve/serve_agent.py
Comment (iycheng) [inline dashboard/modules/serve/serve_agent.py:57]: @edoakes should we keep an in memory copy for the data so that we don't need gcs to retrieve them?

Comment (iycheng) [inline dashboard/modules/serve/serve_agent.py:57]: Also synced with bruce we should just consider it as a failure for now.

File: dashboard/consts.py
Comment (shrekris-anyscale) [inline dashboard/consts.py:29]: If this value is in seconds, can we call this variable `GCS_RPC_TIMEOUT_SECONDS` to match the other constants' formats?

File: dashboard/modules/serve/serve_agent.py
Comment (shrekris-anyscale) [inline dashboard/modules/serve/serve_agent.py:192]: This will only find the controller when it's running in [`detached` mode](https://github.com/ray-project/ray/blob/a4434fac7f63b559e2fc5d763a505160dbae3558/python/ray/serve/_private/api.py#L182-L185). @edoakes Is that alright?

@iycheng Is there a reason we're adding this logic to the Serve agent instead of offloading it to the [`get_global_client`](https://github.com/ray-project/ray/blob/e5a8b1dd55817305ac85237cd99d8b6ed23df294/python/ray/serve/context.py#L36) function?

File: dashboard/modules/serve/tests/test_serve_agent_ft.py
Comment (shrekris-anyscale) [inline dashboard/modules/serve/tests/test_serve_agent_ft.py:27]: Can you add a comment explaining what this test does? I'm not sure what `ft` means in `test_deployment_get_ft`.

File: dashboard/modules/serve/tests/test_serve_agent_ft.py
Comment (shrekris-anyscale) [inline dashboard/modules/serve/tests/test_serve_agent_ft.py:53]: Can you add a comment explaining what this test does? I'm not sure what `ft` means in `test_deployment_get_ft`.

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work so far on the changes! I left a few comments.

File: dashboard/consts.py
Comment (iycheng) [reply dashboard/consts.py:29]: Sure. I'll do. I fail to add seconds because in py it seems like it's seconds by default. but i think following the formats are more important.

File: dashboard/modules/serve/serve_agent.py
Comment (iycheng) [reply dashboard/modules/serve/serve_agent.py:192]: The main reason is that get_global_client is not async which potentially will blocks serve agent and serve agent will blocks dashboard agent, and dashboard agent will blocks healthz, so k8s will kill the raylet.

I think there should be a refactoring or improvements of serve api to make them async but I notice it requires a lot of thing to do and probably not fit for 2.0 in both time constraints and complexity adds.

File: dashboard/modules/serve/tests/test_serve_agent_ft.py
Comment (iycheng) [reply dashboard/modules/serve/tests/test_serve_agent_ft.py:27]: let me change the name: fault_tolerance  (gcs failure won't make agent hangs)

**Review by shrekris-anyscale — APPROVED** [review]
Nice work on the updates. I'll approve this PR for now, and I'll defer to @edoakes on the `detached` mode issue.

@edoakes See [this thread](https://github.com/ray-project/ray/pull/27526#discussion_r940545517) for more context. After this PR is merged, the Serve REST API will only work if the controller is running in `detached` mode.

**Comment by edoakes** [comment]
@shrekris-anyscale I think `detached`-only should be fine given that `serve.run` and the CLI/REST API always start in detached mode.

File: ci/ci.sh
Comment (edoakes) [inline ci/ci.sh:138]: is this intentional?

**Review by edoakes — APPROVED** [review]

File: ci/ci.sh
Comment (iycheng) [reply ci/ci.sh:138]: Yes. When I'm doing the testing, I'm not sure, but it seems that without this, the image I've built will be wrong. I hope it's ok just add it here. (also ok with removing it and add it in other prs).

**Pull request merged**
